### PR TITLE
Implement KCL Deaggregation

### PR DESCRIPTION
**Pull request by justinborromeo**
Fixes https://github.com/apache/incubator-druid/issues/6714.  Rather than having users download the Kinesis Client Library jar, this change adds the capability to perform deaggregation on Kinesis Producer Library-aggregated messages without the KCL dependency.  These changes were all written without any knowledge of how KCL implements deaggregation.

Changes:

1) Messages are checked to see if they are aggregations based on whether they have the magic numbers and their checksum is valid (see https://github.com/awslabs/amazon-kinesis-producer/blob/master/aggregation-format.md for protobuf schema).  If so, they are deaggregated using the Google protobuf library.  If not, they are processed in the same manner as before.

2) The deaggregation config value was removed.  The new deaggregation method can handle both aggregated and non-aggregated messages.

3) A maven step was added to generate a Java file that matches the KPL protobuf schema.

I know my doc changes were minimal so if anyone has any suggestions about what else to write about, feel free to lmk.

### CODE REVIEW

File: extensions-core/kinesis-indexing-service/src/main/java/org/apache/druid/indexing/kinesis/KinesisRecordSupplier.java [hunk 1]
```diff
@@ -747,6 +702,40 @@ private String getSequenceNumberInternal(StreamPartition<String> partition, Stri
 
   }
 
+  @VisibleForTesting
+  List<byte[]> deaggregateKinesisRecord(Record kinesisRecord) throws InvalidProtocolBufferException {
```
Comment (justinborromeo) [inline extensions-core/kinesis-indexing-service/src/main/java/org/apache/druid/indexing/kinesis/KinesisRecordSupplier.java]: I moved this function outside of the PartitionResource class so I could test it.  Is there a way to test private class methods?

File: extensions-core/kinesis-indexing-service/pom.xml [hunk 2]
```diff
@@ -20,90 +20,135 @@
 
 <project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
          xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
-    <modelVersion>4.0.0</modelVersion>
+  <modelVersion>4.0.0</modelVersion>
```
Comment (justinborromeo) [inline extensions-core/kinesis-indexing-service/pom.xml]: Oops changing the indent from 4 spaces to 2 spaces (like all of the other poms) borked the diff.  Mb

File: extensions-core/kinesis-indexing-service/pom.xml [hunk 3]
```diff
@@ -20,90 +20,135 @@
...
+              <inputDirectories>
+                <include>src/main/protobuf</include>
+              </inputDirectories>
+            </configuration>
+          </execution>
+        </executions>
+      </plugin>
+    </plugins>
+  </build>
+  <profiles>
```
Comment (justinborromeo) [inline extensions-core/kinesis-indexing-service/pom.xml]: This profile is the only change in this pom file.

File: extensions-core/kinesis-indexing-service/src/main/java/org/apache/druid/indexing/kinesis/KinesisRecordSupplier.java (hunk 1 above)
Comment (gianm) [reply extensions-core/kinesis-indexing-service/src/main/java/org/apache/druid/indexing/kinesis/KinesisRecordSupplier.java]: Not really a good one. That's usually what `@VisibleForTesting` means: "I would have made this private, except I needed to call it in a unit test."

File: extensions-core/kinesis-indexing-service/pom.xml (hunk 2 above)
Comment (gianm) [reply extensions-core/kinesis-indexing-service/pom.xml]: Fwiw, adding `?w=1` to the PR URL (https://github.com/apache/incubator-druid/pull/6917/files?w=1) gets GitHub to show the diff in whitespace-doesn't-matter mode. In this case it makes it clearer what the real changes are.

File: extensions-core/kinesis-indexing-service/src/main/java/org/apache/druid/indexing/kinesis/KinesisRecordSupplier.java [hunk 4]
```diff
@@ -747,6 +706,41 @@ private String getSequenceNumberInternal(StreamPartition<String> partition, Stri
...
+      kinesisRecordData.get(magicNumbers, 0, magicNumbers.length);
+      kinesisRecordData.get(protobufMessage, 0, protobufMessage.length);
+      kinesisRecordData.get(checksum, 0, checksum.length);
+      byte[] messageHash = Hashing.md5().hashBytes(protobufMessage).asBytes();
+
+      //Check magic numbers and checksum value to see if the message is an aggregate
+      if (Arrays.equals(magicNumbers, KPL_AGGREGATE_MAGIC_NUMBERS)
+          && Arrays.equals(messageHash, checksum)) {
+        List<byte[]> data = new ArrayList<>();
+        AggregatedRecordProtos.AggregatedRecord aggregatedRecord = AggregatedRecordProtos.AggregatedRecord.parseFrom(protobufMessage);
```
[the remark this answers is not in the document]
Comment (gianm) [reply extensions-core/kinesis-indexing-service/src/main/java/org/apache/druid/indexing/kinesis/KinesisRecordSupplier.java]: What's the error look like? (The job looks like it was restarted recently.) How do the protobuf tests make this work - I think they use generated sources too?

If those avenues don't turn up anything interesting, maybe we can suppress something.

[the remark this answers is not in the document]
Comment (justinborromeo) [reply extensions-core/kinesis-indexing-service/src/main/java/org/apache/druid/indexing/kinesis/KinesisRecordSupplier.java]: Avoided the TeamCity inspection errors by removing the generation of `AggregatedRecordProtos` at compile time and just using it as a regular class with warnings suppressed.  This matches the implementation of `ProtoTestEvent` in the `protobuf-extensions` tests.

File: extensions-core/kinesis-indexing-service/src/test/resources/base64aggregatedkinesismessage.txt [hunk 5]
```diff
@@ -0,0 +1 @@
...OiIyMDE5LTAxLTI1VDEwOjI2OjM1LjY3MC0wODowMCIsIm15U3RyaW5nIjoic29tZSB0ZXh0IiwibXlJbnQiOjQsIm15RG91YmxlIjozNi45NjY5MDkyODE0NTI0OX0abwgAGmt7InRpbWVzdGFtcCI6IjIwMTktMDEtMjVUMTA6MjY6MzUuNjcwLTA4OjAwIiwibXlTdHJpbmciOiJzb21lIHRleHQiLCJteUludCI6NCwibXlEb3VibGUiOjk2LjgzMzM5MzE4Mjc0NDI3fRpuCAAaansidGltZXN0YW1wIjoiMjAxOS0wMS0yNVQxMDoyNjozNS42NzAtMDg6MDAiLCJteVN0cmluZyI6InNvbWUgdGV4dCIsIm15SW50Ijo0LCJteURvdWJsZSI6NDUuMTQyMDcwMjkxNTA2MX0abggAGmp7InRpbWVzdGFtcCI6IjIwMTktMDEtMjVUMTA6MjY6MzUuNjcxLTA4OjAwIiwibXlTdHJpbmciOiJzb21lIHRleHQiLCJteUludCI6NCwibXlEb3VibGUiOjcwLjg0MjE5MjY3MjM5Mzl9Gm8IABpreyJ0aW1lc3RhbXAiOiIyMDE5LTAxLTI1VDEwOjI2OjM1LjY3MS0wODowMCIsIm15U3RyaW5nIjoic29tZSB0ZXh0IiwibXlJbnQiOjQsIm15RG91YmxlIjo2Ni40NDU2NDU2ODE2MzkyNX0abwgAGmt7InRpbWVzdGFtcCI6IjIwMTktMDEtMjVUMTA6MjY6MzUuNjcxLTA4OjAwIiwibXlTdHJpbmciOiJzb21lIHRleHQiLCJteUludCI6NCwibXlEb3VibGUiOjUuMDY0MzczODg3MzcxNzMxfRpvCAAaa3sidGltZXN0YW1wIjoiMjAxOS0wMS0yNVQxMDoyNjozNS42NzEtMDg6MDAiLCJteVN0cmluZyI6InNvbWUgdGV4dCIsIm15SW50Ijo0LCJteURvdWJsZSI6NjMuNTQ1MjE5MjM3ODc2MDd9Gm8IABpreyJ0aW1lc3RhbXAiOiIyMDE5LTAxLTI1VDEwOjI2OjM1LjY3Mi0wODowMCIsIm15U3RyaW5nIjoic29tZSB0ZXh0IiwibXlJbnQiOjQsIm15RG91YmxlIjo1My4zNDI4ODM3MjcwNDg5M31foE8ypcllp3vZXupl8YPE
```
Comment (justinborromeo) [inline extensions-core/kinesis-indexing-service/src/test/resources/base64aggregatedkinesismessage.txt]: Does this file need a license?  If so, I'll have to add something to the reader code in tests to ignore the header

Comment (jon-wei) [reply extensions-core/kinesis-indexing-service/src/test/resources/base64aggregatedkinesismessage.txt]: It shouldn't, we have a rat exclusion for `<exclude>**/test/resources/**</exclude>`

**Comment by jihoonson** [comment]
I'm untagging milestone since this issue is not necessarily a release blocker. Feel free to let me know if you think this should be.

**Comment by justinborromeo** [comment]
@jihoonson I agree that this shouldn't be a blocker

**Comment by justinborromeo** [comment]
Unnecessary since KCLv2 is Apache 2.0-licensed

**Pull request closed without merging**